# Lab 2: Arrays and Tables

## Due Sunday, January 19, 11:59 pm 

Welcome to lab 2!  This week, we'll learn about arrays, which allow us to store sequences of data, and *tables*, which let us work with multiple arrays of data about the same things. Arrays are covered in [Chapter 5](http://sierra.ucsd.edu/dsc10-book/chapters/05/Sequences.html) of the textbook, and Tables are described in [Chapter 6](http://sierra.ucsd.edu/dsc10-book/chapters/06/Tables.html).

**Please do not use for-loops for any questions in this lab.** If you don't know what a for-loop is, don't worry -- we haven't covered them yet. But if you do know what they are and are wondering why it's not OK to use them, it is because loops in Python are slow, and looping over arrays and tables should usually be avoided.

First, set up the tests and imports by running the cell below.

In [1]:
import math
import babypandas as bpd

from client.api.notebook import Notebook
ok = Notebook('lab.ok')
_ = ok.auth(inline=True)

Assignment: Lab 2: Arrays and Tables
OK, version v1.14.18

Successfully logged in as h9ryu@ucsd.edu


# 1. Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that's the result of multiplying each number in `billions_of_numbers` by .18 (18%).  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**, like a column in an Excel spreadsheet. 

<img src="excel_array.jpg">

## 1.1. Making arrays
You can type in the data that goes in an array yourself, but that's not typically how programs work. Normally, we create arrays by loading them from an external source, like a data file.

First, though, let's learn how to do it the hard way. To begin, we can make a list of numbers by putting them within square brackets:

In [2]:
my_list = [0.125, 4.75, -1.3]
my_list

[0.125, 4.75, -1.3]

Just like `int`, `float`, and `str`, the `list` is a datatype provided by Python. It is very flexible and easy to work with, but it is *slowwww*.

As data scientists, we'll often be working with millions or even billions of numbers. For this, we need something faster than a `list`. Instead of lists, we will use *arrays*. 

Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie" or, if you prefer to pronounce things incorrectly, "NUM-pee").  The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Data scientists, as well as engineers and scientists of all kinds, use `numpy` frequently, and you'll see quite a bit of it if you're a data science major.

In [3]:
import numpy as np

Now, to create an array, call the function `np.array` with a list of numbers.  Run this cell to see an example:

In [4]:
np.array([0.125, 4.75, -1.3])

array([ 0.125,  4.75 , -1.3  ])

Note that you need the square-brackets here. If you were to try running the following code, Python will yell at you because you forgot them:

```
np.array(0.125, 4.75, -1.3)
```

Arrays themselves are also values, just like numbers and strings.  That means you can assign them names or use them as arguments to functions.

**Question 1.1.1.** Make an array containing the numbers 1, 2, and 3, in that order.  Name it `small_numbers`.

In [5]:
#...small_numbers
small_numbers = np.array([1, 2, 3])
small_numbers

array([1, 2, 3])

In [6]:
_ = ok.grade('q1_1_1')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



**Question 1.1.2.** Make an array containing the numbers 0, 1, -1, $\pi$, and $e$, in that order.  Name it `interesting_numbers`.  *Hint:* remember from the last lab that $\pi$ and $e$ are available from the `math` module, which has already been imported.

In [7]:
#...interesting_numbers
interesting_numbers = np.array([0, 1, -1, math.pi, math.e])
interesting_numbers

array([ 0.        ,  1.        , -1.        ,  3.14159265,  2.71828183])

In [8]:
_ = ok.grade('q1_1_2')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



**Question 1.1.3.** Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you print `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely way of saying that the things in the array are strings. In case you're interested, the `U` means that this string is encoded in [unicode](https://en.wikipedia.org/wiki/Unicode).

In [9]:
#...hello_world_components
hello_world_components = np.array(['Hello', ',', ' ', 'world', '!'])
hello_world_components

array(['Hello', ',', ' ', 'world', '!'], dtype='<U5')

In [10]:
_ = ok.grade('q1_1_3')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  `np.arange(start, stop, space)` produces an array with all the numbers starting at `start` and counting up by `space`, stopping before `stop` is reached.

For example, the value of `np.arange(1, 8, 2)` is an array with elements 1, 3, 5, and 7 -- it starts at 1 and counts up by 2, then stops before 8.  In other words, it makes the same array as `np.array([1, 3, 5, 7])`.

`np.arange(4, 9, 1)` is an array with elements 4, 5, 6, 7, and 8.  (It doesn't contain 9 because `np.arange` stops *before* the stop value is reached.)

**Question 1.1.4.** Use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)

In [11]:
#...multiples_of_99
multiples_of_99 = np.arange(0, 10_000, 99)
multiples_of_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

In [12]:
_ = ok.grade('q1_1_4')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



##### Temperature readings
NOAA (the US National Oceanic and Atmospheric Administration) operates weather stations that measure surface temperatures at different sites around the United States.  The hourly readings are [publicly available](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suppose we download all the hourly data from the Sandy Eggo, California site for the month of December 2015.  To analyze the data, we want to know when each reading was taken, but we find that the data don't include the timestamps of the readings (the time at which each one was taken).

However, we know the first reading was taken at the first instant of December 2015 (midnight on December 1st) and each subsequent reading was taken exactly 1 hour after the last.

**Question 1.1.5.** Create an array of the *time, in seconds, since the start of the month* at which each hourly reading was taken.  Name it `collection_times`.

*Hint 1:* There were 31 days in December, which is equivalent to ($31 \times 24$) hours or ($31 \times 24 \times 60 \times 60$) seconds.  So your array should have $31 \times 24$ elements in it.

*Hint 2:* The `len` function works on arrays, too.  If your `collection_times` isn't passing the tests, check its length and make sure it has $31 \times 24$ elements.

In [13]:
#...collection_times
collection_times = np.arange(0, 31 * 24 * 60 * 60, 60*60)
collection_times

array([      0,    3600,    7200,   10800,   14400,   18000,   21600,
         25200,   28800,   32400,   36000,   39600,   43200,   46800,
         50400,   54000,   57600,   61200,   64800,   68400,   72000,
         75600,   79200,   82800,   86400,   90000,   93600,   97200,
        100800,  104400,  108000,  111600,  115200,  118800,  122400,
        126000,  129600,  133200,  136800,  140400,  144000,  147600,
        151200,  154800,  158400,  162000,  165600,  169200,  172800,
        176400,  180000,  183600,  187200,  190800,  194400,  198000,
        201600,  205200,  208800,  212400,  216000,  219600,  223200,
        226800,  230400,  234000,  237600,  241200,  244800,  248400,
        252000,  255600,  259200,  262800,  266400,  270000,  273600,
        277200,  280800,  284400,  288000,  291600,  295200,  298800,
        302400,  306000,  309600,  313200,  316800,  320400,  324000,
        327600,  331200,  334800,  338400,  342000,  345600,  349200,
        352800,  356

In [14]:
_ = ok.grade('q1_1_5')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



## 1.2. Working with single elements of arrays ("indexing")
Let's work with a more interesting dataset.  The next cell creates an array called `population` that includes estimated world populations in every year from **1950** to roughly the present.  (The estimates come from the [US Census Bureau website](http://www.census.gov/population/international/data/worldpop/table_population.php).)

Rather than type in the data manually, we've loaded them from a file on your computer called `world_population.csv`.  You'll learn how to do that next week.

In [15]:
# Don't worry too much about what goes on in this cell.
population = bpd.read_csv("world_population.csv").get("Population").values
population

array([2557628654, 2594939877, 2636772306, 2682053389, 2730228104,
       2782098943, 2835299673, 2891349717, 2948137248, 3000716593,
       3043001508, 3083966929, 3140093217, 3209827882, 3281201306,
       3350425793, 3420677923, 3490333715, 3562313822, 3637159050,
       3712697742, 3790326948, 3866568653, 3942096442, 4016608813,
       4089083233, 4160185010, 4232084578, 4304105753, 4379013942,
       4451362735, 4534410125, 4614566561, 4695736743, 4774569391,
       4856462699, 4940571232, 5027200492, 5114557167, 5201440110,
       5288955934, 5371585922, 5456136278, 5538268316, 5618682132,
       5699202985, 5779440593, 5857972543, 5935213248, 6012074922,
       6088571383, 6165219247, 6242016348, 6318590956, 6395699509,
       6473044732, 6551263534, 6629913759, 6709049780, 6788214394,
       6866332358, 6944055583, 7022349283, 7101027895, 7178722893,
       7256490011])

Here's how we get the first element of `population`, which is the world population in the first year in the dataset, 1950.

In [16]:
population[0]

2557628654

Notice that we use square brackets here. The square brackets signal that we are *accessing* an element of the array. Square brackets in Python are kind of like subscripts in math.

The value of that expression is the number 2557628654 (around 2.5 billion), because that's the first thing in the array `population`.

Notice that we wrote `population[0]`, not `population[1]`, to get the first element.  This is a weird convention in computer science.  0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

Here are some more examples.  In the examples, we've given names to the things we get out of `population`.  Read and run each cell.

In [17]:
# The third element in the array is the population
# in 1952.
population_1952 = population[2]
population_1952

2636772306

In [18]:
# The thirteenth element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population[12]
population_1962

3140093217

In [19]:
# The 66th element is the population in 2015.
population_2015 = population[65]
population_2015

7256490011

In [20]:
# The array has only 66 elements, so this doesn't work.
# (There's no element with 66 other elements before it.)

population_2016 = population[66]
population_2016

# (after running this cell you can place a # before each line above to make sure that it doesn't run again)

IndexError: index 66 is out of bounds for axis 0 with size 66

In [21]:
# Since make_array returns an array, we can call .item(3)
# on its output to get its 4th element, just like we
# "chained" together calls to the method "replace" earlier.
np.array([-1, -3, 4, -2])[3]

-2

**Question 1.2.1.** Set `population_1973` to the world population in 1973, by getting the appropriate element from `population`.

In [22]:
#...population_1973
population_1973 = population[23]
population_1973

3942096442

In [23]:
_ = ok.grade('q1_2_1')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



## 1.3. Doing something to every element of an array
Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

##### Logarithms
Here is one simple question we might ask about world population:

> How big was the population in *orders of magnitude* in each year?

The logarithm function is one way of measuring how big a number is. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We could try to answer our question like this, using the `log10` function from the `math` module and the `item` method you just saw:

In [24]:
import math
population_1950_magnitude = math.log10(population[0])
population_1951_magnitude = math.log10(population[1])
population_1952_magnitude = math.log10(population[2])
population_1953_magnitude = math.log10(population[3])
...

Ellipsis

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `log10` that takes the logarithm of each element of an array.  It takes a single array of numbers as its argument.  It returns an array of the same length, where the first element of the result is the logarithm of the first element of the argument, and so on.

**Question 1.3.1.** Use NumPy's `log10` function (not `math.log10`!) to compute the logarithms of the world population in every year.  Give the result (an array of 66 numbers) the name `population_magnitudes`.  Your code should be very short.

In [25]:
#...population_magnitudes
population_magnitudes = np.log10(population)
population_magnitudes

array([9.40783749, 9.4141273 , 9.42107263, 9.42846742, 9.43619893,
       9.44437257, 9.45259897, 9.46110062, 9.4695477 , 9.47722498,
       9.48330217, 9.48910971, 9.49694254, 9.50648175, 9.51603288,
       9.5251    , 9.53411218, 9.54286695, 9.55173218, 9.56076229,
       9.56968959, 9.57867667, 9.58732573, 9.59572724, 9.60385954,
       9.61162595, 9.61911264, 9.62655434, 9.63388293, 9.64137633,
       9.64849299, 9.6565208 , 9.66413091, 9.67170374, 9.67893421,
       9.68632006, 9.69377717, 9.70132621, 9.70880804, 9.7161236 ,
       9.72336995, 9.73010253, 9.73688521, 9.74337399, 9.74963446,
       9.75581413, 9.7618858 , 9.76774733, 9.77343633, 9.77902438,
       9.7845154 , 9.78994853, 9.7953249 , 9.80062024, 9.80588805,
       9.81110861, 9.81632507, 9.82150788, 9.82666101, 9.83175555,
       9.83672482, 9.84161319, 9.84648243, 9.85132122, 9.85604719,
       9.8607266 ])

In [26]:
_ = ok.grade('q1_3_1')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<img src="array_logarithm.jpg">

This is called *elementwise* application of the function, since it operates separately on each element of the array it's called on.  The textbook's section on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.log10`.

##### Arithmetic
Arithmetic also works elementwise on arrays.  For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [27]:
population_in_billions = population / 1000000000
population_in_billions

array([2.55762865, 2.59493988, 2.63677231, 2.68205339, 2.7302281 ,
       2.78209894, 2.83529967, 2.89134972, 2.94813725, 3.00071659,
       3.04300151, 3.08396693, 3.14009322, 3.20982788, 3.28120131,
       3.35042579, 3.42067792, 3.49033371, 3.56231382, 3.63715905,
       3.71269774, 3.79032695, 3.86656865, 3.94209644, 4.01660881,
       4.08908323, 4.16018501, 4.23208458, 4.30410575, 4.37901394,
       4.45136274, 4.53441012, 4.61456656, 4.69573674, 4.77456939,
       4.8564627 , 4.94057123, 5.02720049, 5.11455717, 5.20144011,
       5.28895593, 5.37158592, 5.45613628, 5.53826832, 5.61868213,
       5.69920299, 5.77944059, 5.85797254, 5.93521325, 6.01207492,
       6.08857138, 6.16521925, 6.24201635, 6.31859096, 6.39569951,
       6.47304473, 6.55126353, 6.62991376, 6.70904978, 6.78821439,
       6.86633236, 6.94405558, 7.02234928, 7.10102789, 7.17872289,
       7.25649001])

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a tip on several restaurant bills at once (in this case just 3):

In [28]:
restaurant_bills = np.array([20.12, 39.90, 31.01])
print("Restaurant bills:\t", restaurant_bills)
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [20.12 39.9  31.01]
Tips:			 [4.024 7.98  6.202]


<img src="array_multiplication.jpg">

**Question 1.3.2.** Suppose the total charge at a restaurant is the original bill plus the tip.  That means we can multiply the original bill by 1.2 to get the total charge.  Compute the total charge for each bill in `restaurant_bills`.

In [29]:
#...total_charges
total_charges = restaurant_bills * 1.2
total_charges

array([24.144, 47.88 , 37.212])

In [30]:
_ = ok.grade('q1_3_2')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



Let's read in some data to use in the next question.

In [31]:
more_restaurant_bills = bpd.read_csv("more_restaurant_bills.csv").get("Bill").values

**Question 1.3.3.** `more_restaurant_bills` contains 100,000 bills!  Compute the total charge for each one.

In [32]:
#...more_total_charges
more_total_charges = more_restaurant_bills * 1.2
more_total_charges

array([20.244, 20.892, 12.216, ..., 19.308, 18.336, 35.664])

In [33]:
_ = ok.grade('q1_3_3')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



The function `sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

**Question 1.3.4.** What was the sum of all the bills in `more_restaurant_bills`, *including tips*?

In [34]:
#...sum_of_bills
sum_of_bills = sum(more_restaurant_bills * 1.2)
sum_of_bills

1795730.0640000193

In [35]:
_ = ok.grade('q1_3_4')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



**Question 1.3.5.** The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or USBs come in powers of 2, like 16 GB, 32 GB, or 64 GB.)  Use `np.arange` and the exponentiation operator `**` to compute the first 30 powers of 2, starting from `2^0`.

In [36]:
#...powers_of_2
powers_of_2 = 2**np.arange(30)
powers_of_2

array([        1,         2,         4,         8,        16,        32,
              64,       128,       256,       512,      1024,      2048,
            4096,      8192,     16384,     32768,     65536,    131072,
          262144,    524288,   1048576,   2097152,   4194304,   8388608,
        16777216,  33554432,  67108864, 134217728, 268435456, 536870912])

In [37]:
_ = ok.grade('q1_3_5')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



# 2. Growth Rates

If you get stuck on this part of the lab you can always reference this part of the textbook for help [growth rate](http://sierra.ucsd.edu/dsc10-book/chapters/03/2/1/Growth.html)

The relationship between two measurements of the same quantity taken at different times is often expressed as a growth rate. For example, the United States federal government employed 2,766,000 people in 2002 and 2,814,000 people in 2012. To compute a growth rate, we must first decide which value to treat as the initial amount. For values over time, the earlier value is a natural choice. 

Set the variable "intial" to the number of federal government employees in 2002. And set the variable "changed" to the number of federal government employess in 2012.

In python you can use the underscore `_` to help you visualize large numbers. So `1000` and `1_000` are equivalent in python.

In [38]:
initial = ...
changed = ...
print("intial: " + str(initial))
print("changed: " + str(changed))

intial: Ellipsis
changed: Ellipsis


Now compute the ten year growth rate of federal government employees. The equation for growth rate is the differance between the final value and the initial value, divided by the initial value. Remember to use paranthesis. 
`(final - initial) / initial`

In [39]:
#...growth_rate
growth_rate = (2_814_000 - 2_766_000)/2_766_000
growth_rate

0.01735357917570499

In [40]:
_ = ok.grade('q2_1')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



A growth rate can be negative, representing a decrease in some value. For example, the number of manufacturing jobs in the US decreased from 15.3 million in 2002 to 11.9 million in 2012. Calculate the ten year growth rate of US manufacturing jobs.

In [41]:
#...manufacturing_growth_rate
manufacturing_growth_rate = (11.9 - 15.3)/15.3
manufacturing_growth_rate

-0.22222222222222224

In [42]:
_ = ok.grade('q2_2')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



An annual growth rate is a growth rate of some quantity over a single year. An annual growth rate of 0.035, accumulated each year for 10 years, gives a much larger ten-year growth rate of 0.41 (or 41%).

In [43]:
ten_year_growth_rate = 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 * 1.035 - 1
ten_year_growth_rate

0.410598760621121

This same computation can be expressed using names and exponents.

In [44]:
annual_growth_rate = 0.035
ten_year_growth_rate = (1 + annual_growth_rate) ** 10 - 1
ten_year_growth_rate

0.410598760621121

**Question 2.3.** If a stock portfolio had an annual growth rate of 10%, what would be its 15 year growth rate?

In [45]:
#...fifteen_year_growth_rate
fifteen_year_growth_rate = (1 + .1) ** 15 - 1
fifteen_year_growth_rate

3.1772481694156562

In [46]:
_ = ok.grade('q2_3')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



Likewise, a ten-year growth rate can be used to compute an equivalent annual growth rate. Below, t is the number of years that have passed between measurements. The following computes the annual growth rate of federal expenditures over the last 10 years.

In [47]:
initial = 2.37
changed = 3.38
t = 10
annual_growth_rate = (changed/initial) ** (1/t) - 1
annual_growth_rate

0.03613617208346853

**Question 2.4.** If you wanted a stock portfolio with a 15 year growth rate of 100%, what would its annual growth rate have to be?

In [48]:
#...annual_growth_rate
annual_growth_rate = 2 ** (1/15) - 1
annual_growth_rate

0.04729412282062673

In [49]:
_ = ok.grade('q2_4')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



# 3.1. Tables (DataFrames)

## 3.1. Introduction

For a collection of things in the world, an array is useful for describing a single attribute of each thing. For example, among the collection of US States, an array could describe the land area of each. Tables extend this idea by describing multiple attributes for each element of a collection.

In most data science applications, we have data about many entities, but we also have several kinds of data about each entity.

For example, in the cell below we have two arrays. The first one contains the world population in each year (as [estimated](http://www.census.gov/population/international/data/worldpop/table_population.php) by the US Census Bureau), and the second contains the years themselves (in order, so the first elements in the population and the years arrays correspond).

In [50]:
population_amounts = bpd.read_csv("world_population.csv").get("Population").values
population_years = np.arange(1950, 2015+1)
print("Population column:", population_amounts)
print("Years column:", population_years)

Population column: [2557628654 2594939877 2636772306 2682053389 2730228104 2782098943
 2835299673 2891349717 2948137248 3000716593 3043001508 3083966929
 3140093217 3209827882 3281201306 3350425793 3420677923 3490333715
 3562313822 3637159050 3712697742 3790326948 3866568653 3942096442
 4016608813 4089083233 4160185010 4232084578 4304105753 4379013942
 4451362735 4534410125 4614566561 4695736743 4774569391 4856462699
 4940571232 5027200492 5114557167 5201440110 5288955934 5371585922
 5456136278 5538268316 5618682132 5699202985 5779440593 5857972543
 5935213248 6012074922 6088571383 6165219247 6242016348 6318590956
 6395699509 6473044732 6551263534 6629913759 6709049780 6788214394
 6866332358 6944055583 7022349283 7101027895 7178722893 7256490011]
Years column: [1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962 1963
 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 1977
 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991
 1992 1993 1994 

Suppose we want to answer this question:

> When did world population cross 6 billion?

You could technically answer this question just from staring at the arrays, but it's a bit convoluted, since you would have to count the position where the population first crossed 6 billion, then find the corresponding element in the years array. In cases like these, it might be easier to put the data into a 2-dimensional table called a *DataFrame*.

Just as `numpy` provides arrays, a popular package called `pandas` provides `DataFrame`s. `pandas` is *the* tool for doing data science in Python. Unfortunately, `pandas` isn't as cute as its name might suggest: it's very complicated and can be somewhat hard to learn.

Instead of using `pandas`, we'll use a package that we've created especially for DSC 10. It is a *subset* of `pandas`, including only the parts that we think are necessary and throwing out all of the rest. Because it is smaller (and cuter), we've called it `babypandas`. 

You can import `babypandas` using the following code:

In [51]:
import babypandas as bpd


The nice thing about `babypandas` is that it is easier to learn *but* every bit of code you write using `babypandas` will work with `pandas`, too. If you're a data science major, or just going to be doing a lot of data analysis in Python, you'll see quite a lot of `pandas` in your future.

The expression below:

- creates an empty DataFrame using the expression `bpd.DataFrame()`,
- assigns two columns to the DataFrame by calling `assign`,
- assigns the result to the name `population`, and finally
- evaluates `population` so that we can see the table.

`Year` and `Population` are column labels that we have chosen. The names `population_amounts` and `years` were assigned above to two arrays of the same length.

In [52]:
population_table = bpd.DataFrame().assign(
    Population=population_amounts,
    Year=population_years
)
population_table

Population  Year
0   2557628654  1950
1   2594939877  1951
2   2636772306  1952
3   2682053389  1953
4   2730228104  1954
..         ...   ...
61  6944055583  2011
62  7022349283  2012
63  7101027895  2013
64  7178722893  2014
65  7256490011  2015

[66 rows x 2 columns]

Now the data are all together in a single table! It's much easier to parse this data--if you need to know what the population was in 1959, for example, you can tell from a single glance. We'll revisit this table later.

**Question 3.1.1.** In the cell below, we've created 2 arrays. Using the steps above, assign `top_10_movies` to a DataFrame that has two columns called "Rating" and "Name", which hold `top_10_movie_ratings` and `top_10_movie_names` respectively.

In [53]:
top_10_movie_ratings = np.array([9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8])
top_10_movie_names = np.array([
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)'
])

In [54]:
#...top_10_movies
top_10_movies = bpd.DataFrame().assign(Name=top_10_movie_names, Rating=top_10_movie_ratings)
top_10_movies

Name  Rating
0                    The Shawshank Redemption (1994)     9.2
1                               The Godfather (1972)     9.2
2                      The Godfather: Part II (1974)     9.0
3                                Pulp Fiction (1994)     8.9
4                            Schindler's List (1993)     8.9
5  The Lord of the Rings: The Return of the King ...     8.9
6                                12 Angry Men (1957)     8.9
7                             The Dark Knight (2008)     8.9
8             Il buono, il brutto, il cattivo (1966)     8.9
9  The Lord of the Rings: The Fellowship of the R...     8.8

In [55]:
_ = ok.grade('q3_1_1')

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



Suppose you want to add your own ratings to this table. The cell below contains your ranking of each movie:

In [56]:
my_ranking = [8, 2, 1, 9, 7, 10, 6, 4, 3, 5]

**Question 3.1.2** You can use the `assign` method to add a column to an already-existing table, too. Create a new DataFrame called `with_ranking` by adding a column named "Ranking" to the table in `top_10_movies`.

In [57]:
#...with_ranking
with_ranking = top_10_movies.assign(Ranking=my_ranking)
with_ranking

Name  Rating  Ranking
0                    The Shawshank Redemption (1994)     9.2        8
1                               The Godfather (1972)     9.2        2
2                      The Godfather: Part II (1974)     9.0        1
3                                Pulp Fiction (1994)     8.9        9
4                            Schindler's List (1993)     8.9        7
5  The Lord of the Rings: The Return of the King ...     8.9       10
6                                12 Angry Men (1957)     8.9        6
7                             The Dark Knight (2008)     8.9        4
8             Il buono, il brutto, il cattivo (1966)     8.9        3
9  The Lord of the Rings: The Fellowship of the R...     8.8        5

In [ ]:
_ = ok.grade('q3_1_2')

## 3.2. Indexes

You may have noticed that the table of population numbers what looks like an extra, unlabeled column on the left with the numbers 0 through 65. In fact, this is not a column, its what we call an *Index*. The index contains the row labels. Whereas the columns of this table are labeled "Population" and "Year", the rows are labeled 0, 1, ..., 65.

By default, `babypandas` doesn't know how to label the rows, and so it just numbers them (starting with 0). Of course, in this case it makes more sense to use the year as a row's label. We can do this by telling `babypandas` to set the `'Year'` column as the index:

In [ ]:
population_by_year = population_table.set_index('Year')
population_by_year

As we'll see, this does more than make the table look nicer -- it is very useful, too.

**Question 3.2.1** Create a new DataFrame named `top_10_movies_by_name` by taking the DataFrame you made above, `top_10_movies`, and setting the index to be the "Name" column.

In [ ]:
#...top_10_movies_by_name
top_10_movies_by_name = top_10_movies.set_index('Name')
top_10_movies_by_name

In [ ]:
_ = ok.grade('q3_2_1')

You can get an array of row names using `.index`. For instance, the array of row names of the `population` table is:

In [ ]:
population_by_year.index

**Question 3.2.2** Using code, assign to `fourth_movie` the name of the fourth movie in `top_10_movies_by_name`.

In [ ]:
#...fourth_movie
fourth_movie = top_10_movies_by_name.index[3]
fourth_movie

In [ ]:
_ = ok.grade('q3_2_2')

## 3.3 Reading a table from a file
In most cases, we aren't going to go through the trouble of typing in all the data manually. Instead, we can use functions provided by `babypandas`.

The `bpd.read_csv()` function takes one argument, a path to a data file (a string), and returns a `DataFrame`.  There are many formats for data files, but CSV ("comma-separated values") is the most common.

**Question 3.3.1.** The file `imdb.csv` contains a table of information about the 250 highest-rated movies on IMDb.  Load it as a table called `imdb`.

In [ ]:
#...imdb
imdb = bpd.read_csv('imdb.csv')
imdb

In [ ]:
_ = ok.grade('q3_3_1')

Notice the dots in the middle of the table. This means that a lot of the rows have been omitted. This table is big enough that only a few of its rows are displayed, but the others are still there.  10 are shown, so there are 250 movies total.

Where did `imdb.csv` come from? Take a look at [this lab's folder](./). You should see a file called `imdb.csv`.

Open up the `imdb.csv` file in that folder and look at the format. What do you notice? The `.csv` filename ending says that this file is in the [CSV (comma-separated value) format](http://edoceo.com/utilitas/csv-file-format).

**Question 3.3.2.** It makes more sense to use the movie title as the row label. Create a new DataFrame called `imdb_by_name` which uses the movie title as the index.

In [ ]:
#...imdb_by_name
imdb_by_name = imdb.set_index('Title')
imdb_by_name

In [ ]:
_ = ok.grade('q3_3_2')

## 3.4. Series



Suppose we're interested primarily in movie ratings. To extract just this column from the table, we use the `.get` method:

In [ ]:
ratings = imdb_by_name.get('Rating')
ratings

Notice how not only the movie ratings have been returned, but also the name of the movie! This is precisely because we have set the movie title to be the index! For example, if we had asked for the `Rating` column of the original DataFrame, `imdb`, we would see:

In [ ]:
imdb.get('Rating')

This is one way in which indices are very useful.

At first glance, it might look like asking for a column using `.get` returns a table with one column, but that's not quite right. Instead, it returns a special type of thing called a *Series*:

In [ ]:
type(imdb_by_name.get('Rating'))

You can think of a `Series` as an array with an index. Whereas arrays are simple sequences of numbers without labels, `Series` can have labels. This is often very useful.

`ratings` is now a `Series` which contains the column of movie ratings. Suppose we're interested in the rating of a particular movie: `Alien`. To do so, we will use the `.loc` *accessor* which pulls a value from the Series at a particular *loc*ation:

In [ ]:
ratings.loc['Alien']

There are a couple of things to note here. First, those are square brackets around `Alien`. This is because `.loc` is not a function, but an *accessor*. The square brackets signal that we're going to ber extracting an element from the `Series`. Second, we passed in the label as a string.

**Question 3.4.1.** Find the rating of "The Bourne Ultimatum".

In [ ]:
#...bourne_rating
bourne_rating = ratings.loc['The Bourne Ultimatum']
bourne_rating

In [ ]:
_ = ok.grade('q3_4_1')

Now suppose we wanted to know the year in which `Alien` was released. We could do this by getting the column of years:

In [ ]:
years = imdb_by_name.get('Year')
years

And then using `.loc` to get the right entry:

In [ ]:
years.loc['Alien']

We could also do this in one step by *chaining* the operations together:

In [ ]:
imdb_by_name.get('Year').loc['Alien']

This works because Python first evaluates `imdb_by_name.get('Year')` to a Series. It then evaluates the `.loc['Alien']` to return the year.

Chaining is used pretty frequently and can be handy. Just be sure not to chain *too* many things together that your code gets hard to read. You can always save an intermediate result to a variable.

**Question 3.4.2** Find the decade in which "A Beautiful Mind" was released using chaining. Hint: `imbd_by_name` has a column named "Decade".

In [ ]:
#...decade
decade = imdb_by_name.get('Decade').loc['Alien']
decade

In [ ]:
_ = ok.grade('q3_4_2')

## 4. Analyzing datasets
With just a few table methods, we can answer some interesting questions about the IMDb dataset.

If we want just the ratings of the movies, we can get an array that contains the data in that column:

In [ ]:
ratings = imdb_by_name.get("Rating")
ratings

Remember that `ratings` is a `Series`. `Series` objects have some useful methods.

**Question 4.1.** Find the rating of the highest-rated movie in the dataset.

*Hint:* Type `ratings.` and hit Tab to see a list of the available methods. Is there one that looks useful?

In [ ]:
#...highest_rating
highest_rating = ratings.max()
highest_rating

In [ ]:
_ = ok.grade('q4_1')

You probably want to know the *name* of the movie whose rating you found!  To do that, we can sort the whole Series using the `.sort_values` method:

In [ ]:
ratings.sort_values()

So there are actually 2 highest-rated movies in the dataset: *The Shawshank Redemption* and *The Godfather*.

Notice that we are sorting by the ratings, not the labels! Moreover, the label follows the rating as it is sorted. This is exactly what we want.

Had we wanted the highest rating movies on top, we would need to specify that the sorting should not be in ascending order with a *keyword argument*:


In [ ]:
ratings.sort_values(ascending=False)

Not only can we sort `Series`, but we can sort entire `DataFrame`s, too. When we do that, we have to specify the column to sort by:

In [ ]:
imdb_by_name.sort_values('Rating')

Similarly, we can specify that the sort should be in descending order:

In [ ]:
imdb_by_name.sort_values('Rating', ascending=False)


Some details about sorting:

1. The first argument to `sort_values` is the name of a column to sort by.
2. If the column has strings in it, `sort` will sort alphabetically; if the column has numbers, it will sort numerically.
3. The value of `imdb_by_name.sort_values("Rating")` is a *copy*; the `imdb_by_name` table doesn't get modified. For example, if we called `imdb_by_name.sort("Rating")`, then running `imdb_by_name` by itself would still return the unsorted table. To save the result, you should assign it to a new variable.
4. Rows always stick together when a table is sorted.  It wouldn't make sense to sort just one column and leave the other columns alone.  For example, in this case, if we sorted just the `"Rating"` column, the movies would all end up with the wrong ratings.

**Question 4.2.** Create a version of `imdb_by_name` that's sorted chronologically, with the earliest movies first.  Call it `imdb_sorted`.

In [ ]:
#...imdb_sorted
imdb_sorted = imdb_by_name.sort_values('Year')
imdb_sorted

In [ ]:
_ = ok.grade('q4_2')

**Question 4.3.** What's the title of the earliest movie in the dataset?  You could just look this up from the output of the previous cell.  Instead, write Python code to find out.

*Hint:* Remember that the index is an array.

In [ ]:
#...earliest_movie_title
earliest_movie_title = imdb_sorted.index[0]
earliest_movie_title

In [ ]:
_ = ok.grade('q4_3')

Suppose we want to get the rating of the oldest movie in the table. One way to do this is to first find the index label of the oldest movie (which we've already done). We then extract the 'Rating' column and use `.loc` to find the rating of the oldest movie.

In [ ]:
imdb_sorted.get('Rating').loc[earliest_movie_title]

There's a faster way, though. A Series not only has a `.loc` method, but also an `.iloc` method. While `.loc` looks up things by *label*, `.iloc` looks up elements by *integer position*.

Let's remember what is in the "Rating" column:

In [ ]:
imdb_sorted.get('Rating')

If we want the rating of the first row, we can use `.iloc[0]`:

In [ ]:
imdb_sorted.get('Rating').iloc[0]

This returns the exact same thing as `imdb_sorted.get('Rating').loc['The Kid']`; these are two ways of doing the same thing. Usually it is more convenient to access an element by its label rather than by its integer position.

**Question 4.4.** What is the rating of the third oldest movie in the dataset? You could just look this up from the output of the previous cell. Instead, write Python code to find out.

In [ ]:
#...earliest_movie_rating
earliest_movie_rating = imdb_sorted.get('Rating').iloc[2]
earliest_movie_rating

In [ ]:
_ = ok.grade('q4_4')

# 5. Finding pieces of a dataset

Suppose you're interested in movies from the 1950s.  Sorting the table by year doesn't help you, because the 1950s are in the middle of the dataset. Instead, we'll use a feature of `Series` which allows us to easily compare each element in a column to a particular value.

First remember that we can use `.get` to extract a single column. The result is not a `DataFrame`, but rather a `Series`:

In [ ]:
imdb_by_name.get('Decade')

We want to check whether each movie is released in the decade 1940. Python gives us a way of checking whether two things are equal with `==` (remember that `=` is already taken: it assigns values to variable names):

In [ ]:
3 == 4

In [ ]:
3 == 3

`True` and `False` are instances of a type that we haven't seen before:

In [ ]:
type(True)

`bool` stands for "Boolean". We say that "True" and "False" are *Boolean* values.

It turns out that we can easily check if *all* of the elements in a `Series` are equal to something:

In [ ]:
imdb_by_name.get('Decade') == 1950

We see that the result is a new series which has `True` only where the decade was 1950, and `False` everywhere else. We say that the resulting series is a series of *Booleans*, or a *Boolean Series*.

Let's call this result `is_from_fifties`. Its name can be read like it is a question: "is this movie from the 1950s"?

In [ ]:
is_from_1950s = imdb_by_name.get('Decade') == 1950
is_from_1950s

Each row is an answer to the question. Is "The Elephant Man" from the fifties? `False`. Is "All About Eve" from the fifties? `True`.

We can use `is_from_1950s` to select only the rows from `imdb_by_name` for which the answer is `True`. The syntax for this is:

In [ ]:
imdb_by_name[is_from_1950s]

What `imdb_by_name[is_from_1950s]` does, precisely, is to go through the table `imdb_by_name` row by row. If the row named "Singin' in the Rain" has the value `True` in `is_from_1950s`, that row is kept. If the value is `False`, the row is tossed. And so on, for every row.

Note that we could have accomplished this without ever creating the variable `is_from_1950s` by simply placing the code that we used to create the boolean series directly inside the `[...]`:

In [ ]:
imdb_by_name[imdb_by_name.get('Decade') == 1950]

**Question 5.1.** Create a table called `ninety_nine` containing the movies that came out in 1999.

In [ ]:
#...ninety_nine
ninety_nine = imdb_by_name[imdb_by_name.get('Year') == 1999]
ninety_nine

In [ ]:
_ = ok.grade('q5_1')

So far we've only been finding where a column is *exactly* equal to a certain value. However, there are many other comparison operators.  Here are a few:

|Operator|Tests|
|-|-|
|`==`|thing on left is equal to thing on right|
|`!=`|thing on left is *not* equal to thing on right|
|`>`|thing on left is greater than (and not equal to) thing on right|
|`>=`|thing on left is greater than or equal to thing on right|
|`<`|thing on left is less than (and not equal to) thing on right|

The textbook section on selecting rows has more examples.


**Question 5.2.** Using operators from the table above, find all the movies with a rating higher than 8.5.  Put their data in a table called `really_highly_rated`.

In [ ]:
#...really_highly_rated
really_highly_rated = imdb_by_name[imdb_by_name.get('Rating') > 8.5]
really_highly_rated

In [ ]:
_ = ok.grade('q5_2')

What is the highest rating of any movie from the 1920s? We now have the tools to answer questions like these. Breaking it into pieces, we first find all of the movies from the 1920s:

In [ ]:
is_from_1920s = imdb_by_name.get('Decade') == 1920
is_from_1920s

We then select only these movies from our table:

In [ ]:
from_1920s = imdb_by_name[is_from_1920s]
from_1920s

We then find the highest rating out of just these movies:

In [ ]:
from_1920s.get('Rating').max()

Or, if we wanted to do all of this more concisely using chaining:

In [ ]:
imdb_by_name[imdb_by_name.get('Decade') == 1920].get('Rating').max()

**Question 5.3.** Find the average rating for movies released in the 20th century and the average rating for movies released in the 21st century for the movies in `imdb`.

*Hint*: `Series` have a `.mean()` method. Note that the year 2000 is in the 20th century!

In [ ]:
#...average_20th_century_rating
average_20th_century_rating = imdb_by_name[imdb_by_name.get('Year') <= 2000].get('Rating').mean()
average_20th_century_rating

In [ ]:
#...average_21st_century_rating
average_21st_century_rating = imdb_by_name[imdb_by_name.get('Year') > 2000].get('Rating').mean()
average_21st_century_rating

In [ ]:
_ = ok.grade('q5_3')

The property `shape` tells you how many rows and columns are in a table.  (A "property" is just a method that doesn't need to be called by adding parentheses.)

In [ ]:
imdb_by_name.shape

Like an array, you can get the first element of the shape using [0], and the second element using [1]. For instance, the number of rows in `imdb_by_name` is:

In [ ]:
imdb_by_name.shape[0]

We can use this to answer "How many movies are from the 20th century?":

In [ ]:
imdb_by_name[imdb_by_name.get('Year') <= 2000].shape[0]

**Question 5.4.** Use `shape` (and arithmetic) to find the *proportion* of movies in the dataset that were released in the 20th century, and the proportion from the 21st century.

*Hint:* The *proportion* of movies released in the 20th century is the *number* of movies released in the 20th century, divided by the *total number* of movies.

In [ ]:
#...proportion_in_20th_century
proportion_in_20th_century = imdb_by_name[imdb_by_name.get('Year') <= 2000].shape[0] / imdb_by_name.shape[0]
proportion_in_20th_century

In [ ]:
#...proportion_in_21st_century
proportion_in_21st_century = imdb_by_name[imdb_by_name.get('Year') > 2000].shape[0] / imdb_by_name.shape[0]
proportion_in_21st_century

In [ ]:
_ = ok.grade('q5_4')

**Question 5.5.** Check out the `population_by_year` table from the introduction to this lab.  Compute the year when the world population first went above 6 billion.

In [ ]:
#...year_population_crossed_6_billion
year_population_crossed_6_billion = population_by_year[population_by_year.get('Population') > 6_000_000_000].index[0]
year_population_crossed_6_billion

In [ ]:
_ = ok.grade('q5_5')

# Finish Line

Congratulations, you're done with the lab!  Be sure to

- **Verify that all tests pass** (the next cell has a shortcut for that), 
- **Save and Checkpoint** from the `File` menu,
- **Run the last cell to submit your work**

In [ ]:
# For your convenience, you can run this cell to run all the tests at once!
import os
_ = [ok.grade(q[:-3]) for q in os.listdir("tests") if q.startswith('q')]

**Important.** Before you stop working, run this final cell to submit your work.

**If you submit multiple times, your last submission will be counted.**

In [ ]:
_ = ok.submit()